### ISSUE TITLE
Roslyn should infer discard's type from assignment source and shouldn't warn about mismatching nullability with a discard

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#36503]
**Version Used**: 

Branch master (15 Jun 2019)
Latest commit 34d366 by Jason Malinowski:
Merge pull request #36345 from jasonmalinowski/unskip-test

Re-enable MSBuildWorkspaceTests.TestEditorConfigDiscovery

**Steps to Reproduce**:

```cs
#nullable enable
interface I<in T> {}

class C
{
  void M4(I<I<string?>> i1, I<I<string>?> i2)  
  {
    _ = i2 ?? i1; // incorrect warning
    I<I<string>?> x = i2 ?? i1; // correct warning. This is the inferred type of `_` at the line above according to the warning message
    I<I<string?>> y = i2 ?? i1; // this line proves that i2 and i1 doesn't have warnings from conversions in null coalescing expression
    var z = i2 ?? i1; // this line proves that type inference works correctly with 'var' but not with discards

    var q = (_ = i2 ?? i1); // inferring incorrect type for `_ = i2 ?? i1` changes type of `q`
    q = z; // and makes nullability of `q` and `z` incompatible
  }
}
```

**Expected Behavior**:
No warnings for the line `_ = i2 ?? i1;`

**Actual Behavior**:
`warning CS8619: Nullability of reference types in value of type 'I<I<string?>>' doesn't match target type 'I<I<string>?>'.` is reported for `_ = i2 ?? i1;`

**Notes**
I've explained the problem in comments to the original code.
There's no problems converting `i1` or `i2` to the result type of a null coalescing expression as proven by the lines 3 and 4. The inferred result type of the discard is incorrect, it should be the same as `var`.

I believe discards should always have its type equal to the source of assignment. Otherwise adding a discard will change the result type as demonstrated by the variable `q` in the code above

**Comment by jcouv** [comment]
Fixed in 16.2p4. Thanks for reporting this!
We still have some known issues tracking the same problem with discards in deconstruction.
